Dockerfile

```
@@ -1,29 +1,75 @@
-# Stage 1: Build golang dependencies and binaries
-FROM ubuntu:25.04 AS build
+# syntax=docker/dockerfile:1.7.0
 
-ARG VM_DEBUG
+# --- Builder stage ---
+FROM golang:1.24-bookworm AS builder
 
+ARG RUST_VERSION=1.86.0
 
-RUN apt-get -qq update && \
-    apt-get -qq install curl build-essential git golang upx-ucl libjemalloc-dev libjemalloc2 libbz2-dev
-RUN curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh -s -- -q -y
+# Install system dependencies
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    curl gcc make git build-essential \
+    libjemalloc-dev libbz2-dev pkg-config \
+    && apt-get clean && rm -rf /var/lib/apt/lists/*
+
+# Install Rust
+RUN curl https://sh.rustup.rs -sSf | bash -s -- -y --default-toolchain ${RUST_VERSION} --profile minimal
+ENV PATH="/root/.cargo/bin:${PATH}"
 
 WORKDIR /app
 
-# Copy source code
-COPY . .
+# Copy Cargo manifests for dependency caching
+COPY starknet/compiler/rust/Cargo.toml starknet/compiler/rust/Cargo.lock starknet/compiler/rust/
+COPY vm/rust/Cargo.toml vm/rust/Cargo.lock vm/rust/
+COPY core/rust/Cargo.toml core/rust/Cargo.lock core/rust/
+
+# Touch empty lib.rs to satisfy Cargo
+RUN mkdir -p \
+    starknet/compiler/rust/src \
+    vm/rust/src \
+    core/rust/src && \
+    touch starknet/compiler/rust/src/lib.rs \
+            vm/rust/src/lib.rs \
+            core/rust/src/lib.rs
+
+# Pre-fetch dependencies
+RUN cargo fetch --manifest-path=starknet/compiler/rust/Cargo.toml
+RUN cargo fetch --manifest-path=vm/rust/Cargo.toml
+RUN cargo fetch --manifest-path=core/rust/Cargo.toml
```

```suggestion
RUN cargo build --manifest-path=starknet/compiler/rust/Cargo.toml
RUN cargo build --manifest-path=vm/rust/Cargo.toml
RUN cargo build --manifest-path=core/rust/Cargo.toml
```
This is even more aggressive caching, but I think it's optional in general.